src/python/pants/backend/jvm/tasks/jvm_compile/jvm_compile.py

```diff
@@ -395,8 +396,8 @@ def execute(self):
         for ccs in compile_contexts.values():
           cc = self.select_runtime_context(ccs)
           for conf in self._confs:
-            classpath_product.remove_for_target(cc.target, [(conf, cc.classes_dir)])
-            classpath_product.add_for_target(cc.target, [(conf, cc.jar_file)])
+            classpath_product.remove_for_target(cc.target, [(conf, cc.classes_dir.path)])
```

Comment:
I suspect this code doesn't actually remove any more, because it removes the passed value from the set, rather than doing a path check; I think `remove_for_target` needs to iterate across the products, and remove the one with the matching path...

Reply:
Did this in the body of `remove_for_target`.